# Assignment 5, Question 4: Data Exploration

**Points: 15**

In this notebook, you'll explore the clinical trial dataset using pandas selection and filtering techniques.

You'll use utility functions from `q3_data_utils` where helpful, but also demonstrate direct pandas operations.

## Setup

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Import utilities from Q3
from q3_data_utils import load_data, detect_missing, filter_data

# Load the data
df = load_data('data/clinical_trial_raw.csv')
print(f"Loaded {len(df)} patients with {len(df.columns)} variables")

# Prewritten visualization functions for exploration
def plot_value_counts(series, title, figsize=(10, 6)):
    """
    Create a bar chart of value counts.
    
    Args:
        series: pandas Series with value counts
        title: Chart title
        figsize: Figure size tuple
    """
    plt.figure(figsize=figsize)
    series.plot(kind='bar')
    plt.title(title)
    plt.xticks(rotation=45)
    plt.tight_layout()
    plt.show()

def plot_crosstab(crosstab_data, title, figsize=(10, 6)):
    """
    Create a heatmap of crosstab data.
    
    Args:
        crosstab_data: pandas DataFrame from pd.crosstab()
        title: Chart title
        figsize: Figure size tuple
    """
    plt.figure(figsize=figsize)
    plt.imshow(crosstab_data.values, cmap='Blues', aspect='auto')
    plt.colorbar()
    plt.title(title)
    plt.xticks(range(len(crosstab_data.columns)), crosstab_data.columns, rotation=45)
    plt.yticks(range(len(crosstab_data.index)), crosstab_data.index)
    plt.tight_layout()
    plt.show()

Loaded 10000 patients with 18 variables


## Part 1: Basic Exploration (3 points)

Display:
1. Dataset shape
2. Column names and types
3. First 10 rows
4. Summary statistics (.describe())

In [11]:
# TODO: Display dataset info
print("Dataset Info:")
print(df.info())
print("\nFirst 10 rows of the dataset:")
print(df.head(10))
print("Stats on numerical columns:")
print(df.describe())


Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 18 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   patient_id          10000 non-null  object 
 1   age                 10000 non-null  int64  
 2   sex                 10000 non-null  object 
 3   bmi                 9562 non-null   float64
 4   enrollment_date     10000 non-null  object 
 5   systolic_bp         9586 non-null   float64
 6   diastolic_bp        9586 non-null   float64
 7   cholesterol_total   9446 non-null   float64
 8   cholesterol_hdl     9446 non-null   float64
 9   cholesterol_ldl     9446 non-null   float64
 10  glucose_fasting     9631 non-null   float64
 11  site                10000 non-null  object 
 12  intervention_group  10000 non-null  object 
 13  follow_up_months    10000 non-null  int64  
 14  adverse_events      10000 non-null  int64  
 15  outcome_cvd         10000 non-null  obje

## Part 2: Column Selection (3 points)

Demonstrate different selection methods:

1. Select only numeric columns using `.select_dtypes()`
2. Select specific columns by name
3. Select a subset of rows and columns using `.loc[]`

In [ ]:
# TODO: Select numeric columns
numeric_cols = df.select_dtypes(include=['number'])
display(numeric_cols.columns)


In [ ]:
# TODO: Select specific columns
selected_cols = df[['age', 'bmi', 'cholesterol_level', 'treatment_group']]
display(selected_cols.head(10))


In [15]:
# TODO: Use .loc[] to select subset
subset_df = df.loc[0:4,'age':'bmi']
display(subset_df)


age         sex   bmi
0   80           F  29.3
1   80    Female     NaN
2   82      Female  -1.0
3   95      Female  25.4
4   95           M   NaN

## Part 3: Filtering (4 points)

Filter the data to answer these questions:

1. How many patients are over 65 years old?
2. How many patients have systolic BP > 140?
3. Find patients who are both over 65 AND have systolic BP > 140
4. Find patients from Site A or Site B using `.isin()`

In [16]:
# TODO: Filter and count patients over 65
# 1. Use the filter_data utility from Q3
# 2. Create a filter for age > 65
# 3. Apply the filter and count the results
# print(f"Patients over 65: {len(patients_over_65)}")
filter1 = [{'column' : 'age', 'condition' : 'greater_than', 'value' : 65}]
patients_over_65 = filter_data(df, filter1)
print(f"Patients over 65: {patients_over_65.shape[0]}")

Patients over 65: 8326


In [17]:
# TODO: Filter for high BP
# 1. Use the filter_data utility from Q3
# 2. Create a filter for systolic_bp > 140
# 3. Apply the filter and count the results
# print(f"Patients with high BP: {len(high_bp)}")
filter2 = [{'column' : 'systolic_bp', 'condition' : 'greater_than', 'value' : 140}]
high_bp = filter_data(df, filter2)
print(f"Patients with high BP: {high_bp.shape[0]}")


Patients with high BP: 538


In [19]:
# TODO: Multiple conditions with &
# 1. Use filter_data for multiple conditions:
# 2. Create filters for both conditions:
#     {'column': 'age', 'condition': 'greater_than', 'value': 65},
#     {'column': 'systolic_bp', 'condition': 'greater_than', 'value': 140}
# ]
# 3. Apply the filter and count the results
# print(f"Patients over 65 AND high BP: {len(both_conditions)}")
#
# 5. Alternative: Use in_range for age range:
# 5. Create filter for age range 65-100
# 6. Apply the filter and count the results

# me: the filter argument is a list of dictionaries (iterative) so using it is equivalent to & for multiple conditions
filter3 = [
    {'column': 'age', 'condition': 'greater_than', 'value': 65},
    {'column': 'systolic_bp', 'condition': 'greater_than', 'value': 140}
]
both_conditions = filter_data(df, filter3)
print(f"Patients over 65 AND high BP: {both_conditions.shape[0]}")

Patients over 65 AND high BP: 464


In [21]:
# TODO: Filter by site using .isin()
# 1. Use the filter_data utility from Q3
# 2. Create a filter for Site A or Site B
# 3. Apply the filter and count the results
# print(f"Patients from Site A or B: {len(site_ab)}")
filter4 = [{'column' : 'site', 'condition' : 'in_list', 'value' : ['Site A', 'Site B']}]
site_ab = filter_data(df, filter4)
print(f"Patients from Site A or B: {site_ab.shape[0]}")


Patients from Site A or B: 1397


## Part 4: Value Counts and Grouping (5 points)

1. Get value counts for the 'site' column
2. Get value counts for the 'intervention_group' column  
3. Create a crosstab of site vs intervention_group
4. Calculate mean age by site
5. Save the site value counts to `output/q4_site_counts.csv`

In [39]:
# TODO: Value counts and analysis
df['site'] = df['site'].str.lower().str.replace(r"site|_", " ", regex=True).str.strip()
site_counts = df['site'].value_counts()
df['intervention_group'] = df['intervention_group'].str.lower().str.replace("contrl", "control").str.replace(r"treatment|treatmen", " ", regex=True).str.strip()
intervention_counts = df['intervention_group'].value_counts().sort_index()
print("Site counts:", site_counts)
print("\nIntervention counts:", intervention_counts)

site_vs_treatment = df.groupby('site')['intervention_group'].value_counts().unstack()
print("\nSite vs Treatment Group counts:")
display(site_vs_treatment)

age_by_site = df.groupby('site')['age'].mean()
print("\nAverage age by site:", age_by_site)

Site counts: site
a    2956
b    2453
c    2073
d    1501
e    1017
Name: count, dtype: int64

Intervention counts: intervention_group
a          3321
b          3342
control    3337
Name: count, dtype: int64

Site vs Treatment Group counts:


intervention_group    a    b  control
site                                 
a                   964  969     1023
b                   834  839      780
c                   670  694      709
d                   522  489      490
e                   331  351      335


Average age by site: site
a    56.364005
b    62.401957
c    57.312108
d    63.491006
e    57.064897
Name: age, dtype: float64


In [40]:
# TODO: Save output
# site_counts.to_csv('output/q4_site_counts.csv')
site_counts.to_csv('output/q4_site_counts.csv')


## Summary

Write 2-3 sentences about what you learned from exploring this dataset.

**Your summary here:**

TODO: Write your observations


### This is a dataset of 10,000 observations (patients) and 17 variables. of the 538 patients with systolic blood pressure higher than 140, 464 of them are above 65 years. The distribution of treatment groups are similar between the different sites, although some sites had more total data than others.